# **Project 2: Oil and Gas Summary Production Data (1967-1999)**


####Data source site: https://catalog.data.gov/dataset?res_format=CSV
####Datasheet link: https://catalog.data.gov/dataset/oil-and-gas-summary-production-data-1967-1999

In [33]:
import pandas as pd
import numpy as np
from altair import Chart, X, Y, Color, Scale
import altair as alt


In [34]:
production_df = pd.read_csv('/content/drive/My Drive/Jupyter Notebook Project/Oil_and_Gas_Summary_Production_Data__1967-1999.csv')
production_df.head()

Production Year Production Date Entered                         Operator  \
0             1995              12/03/1996              Buffalo China, Inc.   
1             1995              03/29/1996           Copper Ridge Oil, Inc.   
2             1994              02/14/1995  White, Walter W. & Christina L.   
3             1994              01/19/1995              Stiegler, Richard M   
4             1995              02/12/1996                Bucher, Charles J   

        County         Town                    Field Producing Formation  \
0         Erie      BUFFALO                  BUFFALO              MEDINA   
1      Steuben   WEST UNION  BEECH HILL-INDEPENDENCE       FULMER VALLEY   
2  Cattaraugus   CARROLLTON                 BRADFORD            BRADFORD   
3         Erie  CHEEKTOWAGA                  BUFFALO              MEDINA   
4  Cattaraugus     ALLEGANY                FIVE MILE            BRADFORD   

   Active Oil Wells  Inactive Oil Wells  Active Gas Wells  Inactive Gas Wells  \
0                 0                   0                 1                   0   
1                28                   0                 0                   0   
2                 3                   0                 0                   0   
3                 0                   0                 1                   0   
4                10                   0                 0                   0   

   Injection Wells  Disposal Wells Self-use Well  Oil Produced, bbl  \
0                0               0           YES                  0   
1                7               0            NO               1229   
2                0               0            NO                462   
3                0               0           YES                  0   
4                0               0            NO                 45   

   Gas Produced, Mcf  Water produced, bbl  Taxable Gas, Mcf Purchaser Codes  \
0                106                    0                 0             NaN   
1                  0                  180                 0              OA   
2                  0                    0                 0              OA   
3                530                    0                 0             NaN   
4                  0                   80                 0              OA   

                                   Location  
0      BUFFALO, NY\n(42.887691, -78.879374)  
1   WEST UNION, NY\n(42.094951, -77.725816)  
2   CARROLLTON, NY\n(42.023289, -78.628438)  
3  CHEEKTOWAGA, NY\n(42.907071, -78.754318)  
4     ALLEGANY, NY\n(42.088061, -78.491258)

#**Data Column meanings:**

1. **Production Year**: The year during which the oil and gas are produced. 
2. **Production date entered**: The date of production inputted by the operator.
3.**Operator**: The company or individual that are responsible and have the legal authority to drill wells and undertake the production of hydrocarbons that are found.
4. **County**: Specific region of a state or country where the oil and gas wells are located at.
5. **Town**: The town where the oil and gas wells are located at. 
6. **Field**: The geographical area under which an oil or gas reservoir lies.
7. **Producing formation**: The area where there is an underground rock formation from which oil, gas or water is produced
8. **Active oil wells**: Number of wells that are currently producing oil. 
9. **Inactive oil wells**: Number of wells that have not produced oil in 12 months.
10. **Active gas wells**: Number of wells that are currently producing gas. 
11. **Inactive gas wells**: Number of wells that have not produced gas in 12 months.  
12. **Injection wells**: An injection well is used to place fluid underground into porous geologic formations. This could be steam, carbon dioxide, water or other substances to help maintain reservoir pressure, heat the oil or lower its viscosity, allowing it to flow to a producing well nearby.
13. **Disposal wells**: Number of depleted oil or gas wells, into which waste fluids can be injected for safe disposal (extra info: Water specific to fracture return may be recycled for further use, or disposed by injection into deep subsurface formations, through a disposal well.)
14. **Self-use well**: 
15. **Oil produced (bbl)**: Total number of barrels of oil produced by the active oil wells. 
16. **Gas produced (mcf)**: Amount of gas, in Mille Cubic Feet or 1000 cubic feet,  produced by active gas wells.
17. **Water produced (bbl)**: Total number of barrels of water produced by the active oil and gas wells.
18. **Taxable gas (mcf)**: Amount of gas, in Mille Cubic Feet or 1000 cubic feet, that can be taxed. 
19. **Purchaser codes**: Specifies a code to specify the purchaser who normally handles this vendor's account.
20. **Location**: The country, state, and longitude and latitude 



In [35]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


#**Cleaning the Data:**

####First let's check which columns have empty observations. 

In [36]:
production_df.shape

(30053, 20)

In [37]:
production_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30053 entries, 0 to 30052
Data columns (total 20 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Production Year          30053 non-null  int64 
 1   Production Date Entered  30053 non-null  object
 2   Operator                 30053 non-null  object
 3   County                   30022 non-null  object
 4   Town                     29396 non-null  object
 5   Field                    28772 non-null  object
 6   Producing Formation      29393 non-null  object
 7   Active Oil Wells         30053 non-null  int64 
 8   Inactive Oil Wells       30053 non-null  int64 
 9   Active Gas Wells         30053 non-null  int64 
 10  Inactive Gas Wells       30053 non-null  int64 
 11  Injection Wells          30053 non-null  int64 
 12  Disposal Wells           30053 non-null  int64 
 13  Self-use Well            29434 non-null  object
 14  Oil Produced, bbl        30053 non-nul

####Since there are supposed to be 30,053 observations in total, we can see that the County, Town, Field, Producing Formation, Self-Use Well, and Purchaser Codes columns have empty rows. We will now remove these empty rows. We will not bother about the Self-Use Well and Purchaser Codes columns since those are not used in our analysis. 


In [38]:
production_df = production_df.loc[production_df['County'].notnull()]
production_df = production_df.loc[production_df['Town'].notnull()]
production_df = production_df.loc[production_df['Field'].notnull()]
production_df = production_df.loc[production_df['Producing Formation'].notnull()]
production_df.shape


(28357, 20)

####Since the observations have decreased from 30,053 to 28,357, the null rows have been removed. 

#**Analysis Questions:**


---


###**1) The Renewable Energy movement started during the 1970s. Assuming that this movement would decrease the dependency on natural gas and oil by making people more aware, do the number of active  oil and gas wells decrease over the years? Can we also see the same effect on the number of disposal wells?**


---













#### First, we will make a new dataframe `NR_df` which only has the columns that we need for this analysis. 

In [39]:
NR_df = production_df[['Production Year', 'Active Oil Wells', 'Active Gas Wells', 'Disposal Wells']]
NR_df.head()

Production Year  Active Oil Wells  Active Gas Wells  Disposal Wells
0             1995                 0                 1               0
1             1995                28                 0               0
2             1994                 3                 0               0
3             1994                 0                 1               0
4             1995                10                 0               0

#### We will now group by the production year to see the counts of the 5 different types of wells in each year.

In [40]:
NR_df1 = NR_df.groupby('Production Year').sum()
NR_df1.head()

Active Oil Wells  Active Gas Wells  Disposal Wells
Production Year                                                    
1967                         3250               484               0
1968                         3678               515               0
1969                         4117               481               0
1970                         4261               446               0
1971                         4229               409               0

####Using the `melt()` function we will turn this dataframe from a wide form to a long form since Altair charts work best with the long form.

In [41]:
NR_df1 = NR_df1.reset_index()
NR_df1.head()

Production Year  Active Oil Wells  Active Gas Wells  Disposal Wells
0             1967              3250               484               0
1             1968              3678               515               0
2             1969              4117               481               0
3             1970              4261               446               0
4             1971              4229               409               0

In [42]:
NR_df1_melt = NR_df1.melt(id_vars='Production Year', value_vars=['Active Oil Wells', 'Active Gas Wells', 'Disposal Wells'])
NR_df1_melt

Production Year          variable  value
0              1967  Active Oil Wells   3250
1              1968  Active Oil Wells   3678
2              1969  Active Oil Wells   4117
3              1970  Active Oil Wells   4261
4              1971  Active Oil Wells   4229
..              ...               ...    ...
94             1995    Disposal Wells      0
95             1996    Disposal Wells      1
96             1997    Disposal Wells      5
97             1998    Disposal Wells      0
98             1999    Disposal Wells      0

[99 rows x 3 columns]

####We are now representing this melted dataframe using a line chart. 

In [43]:
Chart(NR_df1_melt).mark_line().encode(x='Production Year:O', y='value', color='variable')

alt.Chart(...)

#### Unfortunately, from the line chart above, we can see that the Renewable Energy movement did not decrease the number of active oil wells and gas wells. Infact, there is an increase in active gas wells significantly from 1975 which is contradictory to the purpose of the movement. We do see a decline in number of active oil wells from 1978 which might indicate more dependency on renewable sources. However, we end up having more number of active oil wells in 1999 than in 1967. For the disposal wells, there seems to be no significant change except for the brief increases in number during 1978, 1987 and 1992. 



---


###**2) Which producing formation produces the highest percentage of taxable gases in each decade? Is there any formation that has been producing highest amount of taxable gas throughout the years?**

---







####We first extract the columns that we need that are Production Year, Producing Formation, Gas Produces, and Taxable Gas. 

In [44]:
TG_df = production_df[['Production Year', 'Producing Formation', 'Gas Produced, Mcf', 'Taxable Gas, Mcf']]
TG_df.head()

Production Year Producing Formation  Gas Produced, Mcf  Taxable Gas, Mcf
0             1995              MEDINA                106                 0
1             1995       FULMER VALLEY                  0                 0
2             1994            BRADFORD                  0                 0
3             1994              MEDINA                530                 0
4             1995            BRADFORD                  0                 0

####To calculate percentage, we have to remove all rows where the Gas Produced is 0, since we do not want an undefined value. 

In [45]:
TG_df = TG_df.loc[TG_df['Gas Produced, Mcf'] > 0]
TG_df

Production Year   Producing Formation  Gas Produced, Mcf  \
0                 1995                MEDINA                106   
3                 1994                MEDINA                530   
5                 1995                MEDINA              27742   
6                 1994  ONONDAGA-BASS ISLAND               1582   
7                 1994                MEDINA                500   
...                ...                   ...                ...   
30043             1994                MEDINA               6745   
30047             1994                MEDINA             203147   
30048             1995                MEDINA              35272   
30051             1995                MEDINA               3370   
30052             1995                MEDINA              30615   

       Taxable Gas, Mcf  
0                     0  
3                     0  
5                 24177  
6                  1439  
7                     0  
...                 ...  
30043              6745  
30047            203147  
30048             24002  
30051              3370  
30052             30615  

[17032 rows x 4 columns]

####First, let's deal with the 1960s.

In [46]:
TG_1960s = TG_df.loc[TG_df['Production Year'] < 1970]
TG_1960s = TG_1960s.sort_values(by='Production Year')
TG_1960s

Production Year Producing Formation  Gas Produced, Mcf  Taxable Gas, Mcf
7945             1967              ONEIDA               1427                 0
8522             1967              MEDINA              35317             35317
8517             1967              MEDINA               1138              1138
8506             1967            ORISKANY               8080              8080
8505             1967              MEDINA               5131              5131
...               ...                 ...                ...               ...
8454             1969              MEDINA              15922             15922
8445             1969              MEDINA                232                 0
8425             1969              MEDINA               7740              7740
8078             1969              MEDINA              13322             13322
8524             1969              MEDINA              20823             20823

[244 rows x 4 columns]

####We will then group by Producing formation and disregard the 'Production Year' column. 

In [47]:
TG_1960s = TG_1960s.groupby('Producing Formation').sum()
TG_1960s = TG_1960s[[ 'Gas Produced, Mcf', 'Taxable Gas, Mcf']]
TG_1960s.head()

Gas Produced, Mcf  Taxable Gas, Mcf
Producing Formation                                     
AKRON                           199018            199018
BRADFORD                         47386             21831
CHEMUNG                         164446            164446
FULMER VALLEY                   210670             24420
GLADE                              440                 0

####We then have to create a new column called 'Percentage of Taxable Gas'. 

In [48]:
TG_1960s['Percentage of Taxable Gas'] = (TG_1960s['Taxable Gas, Mcf']/TG_1960s['Gas Produced, Mcf'] )*100
TG_1960s = TG_1960s.round(2)
TG_1960s.head()

Gas Produced, Mcf  Taxable Gas, Mcf  \
Producing Formation                                        
AKRON                           199018            199018   
BRADFORD                         47386             21831   
CHEMUNG                         164446            164446   
FULMER VALLEY                   210670             24420   
GLADE                              440                 0   

                     Percentage of Taxable Gas  
Producing Formation                             
AKRON                                   100.00  
BRADFORD                                 46.07  
CHEMUNG                                 100.00  
FULMER VALLEY                            11.59  
GLADE                                     0.00

####Let's represent the percentage of taxable gas of each producing formation in a bar graph and give the bar with highest percentage an orange color.

In [49]:
TG_1960s = TG_1960s['Percentage of Taxable Gas']
TG_1960s = TG_1960s.reset_index()
TG_1960s.head()

Producing Formation  Percentage of Taxable Gas
0               AKRON                     100.00
1            BRADFORD                      46.07
2             CHEMUNG                     100.00
3       FULMER VALLEY                      11.59
4               GLADE                       0.00

In [50]:
Chart(TG_1960s).mark_bar().encode(x='Producing Formation:O', y='Percentage of Taxable Gas', 
                                  color=alt.condition(
                                  alt.datum['Percentage of Taxable Gas'] == TG_1960s['Percentage of Taxable Gas'].max(),  
                                  alt.value('orange'),   
                                  alt.value('steelblue')   
                                                     )).properties(width=600)


alt.Chart(...)

####From the bar graph above, we can conclude that Akron, Chemung, Clinton, Hamilton, Marcellus, and Theresa produce the highest percentage of taxable gases in the 1960s

####Let's repeat all steps for the 1970s. 

In [51]:
TG_1970s = TG_df.loc[TG_df['Production Year'] < 1980]
TG_1970s = TG_1970s.groupby('Producing Formation').sum()
TG_1970s = TG_1970s[[ 'Gas Produced, Mcf', 'Taxable Gas, Mcf']]
TG_1970s['Percentage of Taxable Gas'] = (TG_1970s['Taxable Gas, Mcf']/TG_1970s['Gas Produced, Mcf'] )*100
TG_1970s = TG_1970s.round(2)
TG_1970s = TG_1970s['Percentage of Taxable Gas']
TG_1970s = TG_1970s.reset_index()
TG_1970s.head()

Producing Formation  Percentage of Taxable Gas
0               AKRON                      60.87
1            BRADFORD                      74.85
2      BRADFORD THIRD                       0.00
3             CHEMUNG                      74.68
4            CHIPMUNK                      43.63

In [52]:
Chart(TG_1970s).mark_bar().encode(x='Producing Formation:O', y='Percentage of Taxable Gas', 
                                  color=alt.condition(
                                  alt.datum['Percentage of Taxable Gas'] == TG_1970s['Percentage of Taxable Gas'].max(),  
                                  alt.value('orange'),   
                                  alt.value('steelblue')   
                                                     )).properties(width=800)

alt.Chart(...)

####From the bar graph above, we can conclude that Clinton, Hamilton, Herkimer, Onondagaoriskany, Theresa and Trenton produce the highest percentage of taxable gases in the 1970s. 

####Let's repeat all steps for the 1980s. 

In [53]:
TG_1980s = TG_df.loc[TG_df['Production Year'] < 1990]
TG_1980s = TG_1980s.groupby('Producing Formation').sum()
TG_1980s = TG_1980s[[ 'Gas Produced, Mcf', 'Taxable Gas, Mcf']]
TG_1980s['Percentage of Taxable Gas'] = (TG_1980s['Taxable Gas, Mcf']/TG_1980s['Gas Produced, Mcf'] )*100
TG_1980s = TG_1980s.round(2)
TG_1980s = TG_1980s['Percentage of Taxable Gas']
TG_1980s = TG_1980s.reset_index()
TG_1980s.head()

Producing Formation  Percentage of Taxable Gas
0               AKRON                      32.75
1         BASS ISLAND                       0.00
2  BASS ISLAND/MEDINA                       0.00
3         BLACK RIVER                       0.00
4            BRADFORD                      57.83

In [54]:
Chart(TG_1980s).mark_bar().encode(x='Producing Formation:O', y='Percentage of Taxable Gas', 
                                  color=alt.condition(
                                  alt.datum['Percentage of Taxable Gas'] == TG_1980s['Percentage of Taxable Gas'].max(),  
                                  alt.value('orange'),   
                                  alt.value('steelblue')   
                                                     )).properties(width=1000)

alt.Chart(...)

####From the bar graph above, we can conclude that Theresa and Whirlpool produce the highest percentage of taxable gases in the 1980s. 

####Let's repeat all steps for the 1990s. 

In [55]:
TG_1990s = TG_df.loc[TG_df['Production Year'] < 2000]
TG_1990s = TG_1990s.groupby('Producing Formation').sum()
TG_1990s = TG_1990s[[ 'Gas Produced, Mcf', 'Taxable Gas, Mcf']]
TG_1990s['Percentage of Taxable Gas'] = (TG_1990s['Taxable Gas, Mcf']/TG_1990s['Gas Produced, Mcf'] )*100
TG_1990s = TG_1990s.round(2)
TG_1990s = TG_1990s['Percentage of Taxable Gas']
TG_1990s = TG_1990s.reset_index()
TG_1990s.head()

Producing Formation  Percentage of Taxable Gas
0               AKRON                      42.05
1         BASS ISLAND                      13.81
2  BASS ISLAND/MEDINA                       0.00
3         BLACK RIVER                      98.18
4            BRADFORD                      54.31

In [56]:
Chart(TG_1990s).mark_bar().encode(x='Producing Formation:O', y='Percentage of Taxable Gas', 
                                  color=alt.condition(
                                  alt.datum['Percentage of Taxable Gas'] == TG_1990s['Percentage of Taxable Gas'].max(),  
                                  alt.value('orange'),   
                                  alt.value('steelblue')   
                                                     )).properties(width=1200)

alt.Chart(...)

####From the bar graph above, we can conclude that Chipmunk and Bradford 2nd, Littlefalls, NoneSpecified, Sodusshale produce the highest percentage of taxable gases in the 1990s. 

####From the bar graphs above, we saw that there is no single producing formation which created the highest percentage of taxable gases throughout the years. However, we do see Theresa having mazimum production from the 1960s to the 1980s. 



---

##**3) How has the water production changed throughout the decades? To which field has the highest water production shifted throughout the decades?**



---





####First we create a new dataframe constituting the columns we need which are Production year, Field, Water produced.

In [57]:
Field_df = production_df.copy()[['Production Year','Field','Water produced, bbl']]
Field_df.head()

Production Year                    Field  Water produced, bbl
0             1995                  BUFFALO                    0
1             1995  BEECH HILL-INDEPENDENCE                  180
2             1994                 BRADFORD                    0
3             1994                  BUFFALO                    0
4             1995                FIVE MILE                   80

####We now have to create a new column called decades which gives the production years its corresponding decade. First let's create a function to do this.

In [58]:
def decade(Production_Year):
  if Production_Year < 1970:
    return '1960s'
  elif Production_Year < 1980:
    return '1970s'
  elif Production_Year < 1990:
    return '1980s'
  else:
    return '1990s'

####Now, we will apply the `decade` function to the new column called Decade. 





In [59]:
Field_df['Decade'] = Field_df['Production Year'].apply(decade)
Field_df.head()

Production Year                    Field  Water produced, bbl Decade
0             1995                  BUFFALO                    0  1990s
1             1995  BEECH HILL-INDEPENDENCE                  180  1990s
2             1994                 BRADFORD                    0  1990s
3             1994                  BUFFALO                    0  1990s
4             1995                FIVE MILE                   80  1990s

####We will now make a pivot table with the Field column as the index and Decade coulm as the columns. The values are sum of amount of water produced in bbl.

In [60]:
Field_pivot = Field_df.pivot_table(index='Field',
               columns='Decade',
               values='Water produced, bbl',
               aggfunc='sum').fillna(0)
Field_pivot.head()

Decade           1960s  1970s   1980s    1990s
Field                                         
ADRIAN REEF        0.0    0.0     0.0      1.0
ADRIAN STORAGE     0.0    0.0     0.0      0.0
AKRON              0.0    0.0     5.0    438.0
ALDEN-LANCASTER    0.0    0.0  4254.0  12926.0
ALEXANDER          0.0    0.0     0.0    122.0

In [61]:
Field_pivot = Field_pivot.reset_index()
Field_pivot.head()

Decade            Field  1960s  1970s   1980s    1990s
0           ADRIAN REEF    0.0    0.0     0.0      1.0
1        ADRIAN STORAGE    0.0    0.0     0.0      0.0
2                 AKRON    0.0    0.0     5.0    438.0
3       ALDEN-LANCASTER    0.0    0.0  4254.0  12926.0
4             ALEXANDER    0.0    0.0     0.0    122.0

####Let's represent the pivot table in a scatter plot. 

In [62]:
First = Chart(Field_pivot).mark_point(color = 'red').encode(x='Field',
                              y='1960s' ).interactive()
                        
Second = Chart(Field_pivot).mark_point(color = 'green').encode(x='Field',
                              y='1970s').interactive()
 
Third = Chart(Field_pivot).mark_point(color = 'orange').encode(x='Field',
                              y='1980s').interactive()
 
Fourth = Chart(Field_pivot).mark_point(color = 'blue').encode(x='Field',
                              y='1990s').interactive()
First + Second + Third + Fourth


alt.LayerChart(...)

####From the above analysis we can see that for decades of 1960s, 1970s and 1990s Bradford has always had the highest water production. But for the decade of 1980 Chipmunk had the highest water production. 

####Lets see if Bradford and Chipmunk are in the top two Fields that produce the most Oil.

In [63]:
BC_df = production_df[['Field','Gas Produced, Mcf']]
BC_df = BC_df.groupby('Field').sum()
BC_df = BC_df.sort_values(by = 'Gas Produced, Mcf', ascending = False)
BC_df.head()

Gas Produced, Mcf
Field                              
LAKESHORE                 289107481
WEST AUBURN                27428532
STAGECOACH                 14252405
FAYETTE-WATERLOO           14104317
ALDEN-LANCASTER            14070664

####We dont see both Bradford and Chipmunk even in the top 5 which goes against the study that more gas production means higher amount of water produced. 